books/RayTracingTheNextWeek.html

```diff
@@ -2256,7 +2256,319 @@
 
 
 
-Rectangles and Lights
+Quadrilaterals
+====================================================================================================
+We've managed to get more than half way through this three-book series using spheres as our only
+geometric primitive. Time to add our second primitive: the quadrilateral.
+
+
+Ray-Plane Intersection
+-----------------------
+Spheres are generally the first ray tracing primitive because their implicit formula makes it so
+easy to solve for ray intersection. Like spheres, planes also have an implicit formula, and
+ray-plane intersection is even _easier_ to solve.
+
+You may already know this implicit formula for a plane:
+
+  $$ Ax + By + Cz + D = 0 $$
+
+That is, a plane is the set of all points $\mathbf{P} = (x,y,z)$ that satisfy the formula above. It
+makes things slightly easier to use the alternate formulation:
+
+  $$ Ax + By + Cz = D $$
+
+(We didn't flip the sign of D because it's just some constant that we'll figure out later.)
+
+Here's another way to think of this formula: given the plane perpendicular to the normal vector
+$\mathbf{n} = (A,B,C)$, and the position vector $\mathbf{v} = (x,y,z)$ (that is, the vector from the
+origin to any point on the plane), then
+
+  $$ \mathbf{n} \cdot \mathbf{v} = D $$
+
+Now to find the intersection with some ray $\mathbf{R}(t) = \mathbf{P} + t\mathbf{d}$. Plugging in
+the ray equation, we get
+
+  $$ \mathbf{n} \cdot ( \mathbf{P} + t \mathbf{d} ) = D $$
+
+Solving for $t$:
+
+  $$ \mathbf{n} \cdot \mathbf{P} + \mathbf{n} \cdot t \mathbf{d}  = D $$
+
+  $$ \mathbf{n} \cdot \mathbf{P} + t(\mathbf{n} \cdot \mathbf{d}) = D $$
+
+  $$ t = \frac{D - \mathbf{n} \cdot \mathbf{P}}{\mathbf{n} \cdot \mathbf{d}} $$
```

Comment:
We also implicitly check for a positive t below with:

```
            // Return false if the hit point parameter t is outside the ray interval.
            auto t = (-D - dot(normal, r.origin())) / denom;
            if (!ray_t.contains(t))
                return false;
```

This should be called out in the text, and also probably in the comments.

Reply:
It might be worth mentioning something like,

3 possibilities exist:
1. No intersection, zero in denominator
2. A solution where the t value is negative, so no intersection
3. A solution where the t value is positive, so an intersection exists

It's only on case three where we want to continue searching for the actual point of intersection

Reply:
Good point calling out the zero denominator. I don't think we need to comment about negative _t_ values. For one, it's actually _t_ values _less_ than some given value (not necessarily zero), and two, this is already consistent with our hit parameter handling for all other values.

Ah, what the heck, I'll add a sentence.